tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15541,46 +15541,31 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, llvm::StringRef StageName,
       S.Diags.Report(NodeDG->getLocation(), diag::note_defined_here)
           << NodeDG->getSpelling();
     }
-    // Node with NodeMaxDispatchGrid must have SV_DispatchGrid semantic.
-    if (NodeMDG) {
-      bool Found = false;
-      for (FunctionDecl::param_iterator I = FD->param_begin(),
-                                        E = FD->param_end();
-           I != E; ++I) {
-        QualType ParamType = (*I)->getType().getCanonicalType();
-
-        // Find parameter that is the node input record
-        if (hlsl::IsHLSLNodeInputType(ParamType)) {
-          // Node input records are template types
-          if (const RecordType *NodeInputRT = dyn_cast<RecordType>(ParamType)) {
-            if (const ClassTemplateSpecializationDecl *templateDecl =
-                    dyn_cast<ClassTemplateSpecializationDecl>(
-                        NodeInputRT->getDecl())) {
-
-              // Get the input record struct
-              auto &TemplateArgs = templateDecl->getTemplateArgs();
-              DXASSERT_NOMSG(TemplateArgs.size() >= 1);
-              QualType Arg0Type = TemplateArgs.get(0).getAsType();
-              const RecordType *NodeInputStructType =
-                  Arg0Type->getAsStructureType();
-              if (nullptr != NodeInputStructType) {
-                CXXRecordDecl *NodeInputStructDecl =
-                    dyn_cast<CXXRecordDecl>(NodeInputStructType->getDecl());
-                if (nullptr != NodeInputStructDecl) {
-                  // Make sure there is exactly one SV_DispatchGrid semantics
-                  // and it has correct type.
-                  DiagnoseDispatchGridSemantics(S, NodeInputStructDecl, Found);
-                }
-              }
-            }
-          }
+    // Diagnose dispatch grid semantics.
+    bool Found = false;
+    for (ParmVarDecl *PD : FD->params()) {
+      QualType ParamType = PD->getType().getCanonicalType();
+
+      // Find parameter that is the node input record
+      if (hlsl::IsHLSLNodeType(ParamType)) {
+        // Node records are template types
+        if (CXXRecordDecl *NodeStructDecl =
+                hlsl::GetHLSLNodeRecordDecl(ParamType)) {
+          // Diagnose any SV_DispatchGrid semantics used in record.
+          // OutputFound provides the output needed by
+          // DiagnoseDispatchGridSemantics, but isn't used otherwise.
+          bool OutputFound = false;
+          DiagnoseDispatchGridSemantics(
+              S, NodeStructDecl,
+              hlsl::IsHLSLNodeOutputType(ParamType) ? OutputFound : Found);
```

Comment:
I jus realized that this isn't in the right location for diagnosing output nodes, since we are under a broadcasting-only branch (the else branch of `if (NodeLaunchTy != DXIL::NodeLaunchType::Broadcasting)`).

So, we either need to move this param iteration out of this branch, or copy it for outputs after this branch.  If we move this loop out of the branch, for inputs, we only want to check if it's broadcasting, then we have to check `if (NodeMDG && !Found)` only if broadcasting as well.

So maybe an independent loop for checking outputs after this branch would be simpler to read.